File: src/main/java/net/minestom/server/statistic/StatisticType.java
package net.minestom.server.statistic;

import net.kyori.adventure.key.Key;
import net.kyori.adventure.key.KeyPattern;
import net.minestom.server.registry.StaticProtocolObject;
import org.jspecify.annotations.Nullable;

import java.util.Collection;

public sealed interface StatisticType extends StaticProtocolObject<StatisticType>, StatisticTypes permits StatisticTypeImpl {

    static Collection<StatisticType> values() {
        return StatisticTypeImpl.REGISTRY.values();
    }

    static @Nullable StatisticType fromKey(@KeyPattern String key) {
        return fromKey(Key.key(key));
    }

    static @Nullable StatisticType fromKey(Key key) {
        return StatisticTypeImpl.REGISTRY.get(key);
    }

    static @Nullable StatisticType fromId(int id) {
        return StatisticTypeImpl.REGISTRY.get(id);
    }

}

File: src/main/java/net/minestom/server/statistic/StatisticType.kt
package net.minestom.server.statistic

import net.minestom.server.registry.ProtocolObject
import net.minestom.server.statistic.StatisticTypes
import net.minestom.server.statistic.StatisticType
import net.minestom.server.statistic.StatisticTypeImpl
import net.minestom.server.utils.NamespaceID
import net.minestom.server.statistic.StatisticCategory

interface StatisticType : ProtocolObject, StatisticTypes {
    companion object {
        fun values(): Collection<StatisticType?> {
            return StatisticTypeImpl.Companion.values()
        }

        fun fromNamespaceId(namespaceID: String): StatisticType? {
            return StatisticTypeImpl.Companion.getSafe(namespaceID)
        }

        fun fromNamespaceId(namespaceID: NamespaceID): StatisticType? {
            return fromNamespaceId(namespaceID.asString())
        }

        fun fromId(id: Int): StatisticType? {
            return StatisticTypeImpl.Companion.getId(id)
        }
    }
}